src/main/scala/mesosphere/mesos/TaskBuilder.scala

```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
     builder.build()
   }
 
-  def portsEnv(definedPorts: Seq[Int], assignedPorts: Seq[Int], portNames: Seq[Option[String]]): Map[String, String] = {
+  // portsEnv generates $PORT{x} and $PORT_{y} environment variables, wherein `x` is an index into
+  // the portDefinitions or portMappings array and `y` is a non-zero port specifically requested by
+  // the application specification.
+  //
+  // @param definedPorts are either declared container ports (if port mappings are specified) or host ports; may be 0's
+  // @param assignedPorts are "real" host ports allocated from Mesos resource offers
+  // @return a dictionary of variables that should be added to a tasks environment
+  def portsEnv(
+    definedPorts: Seq[Int],
+    assignedPorts: Seq[Option[Int]],
+    portNames: Seq[Option[String]]): Map[String, String] = {
     if (assignedPorts.isEmpty) {
       Map.empty
     }
     else {
       val env = Map.newBuilder[String, String]
+      val generatedPorts = Map.newBuilder[Int, Int] // index -> container port
+
+      object ContainerPortGenerator {
+        // track which port numbers are already referenced by PORT_xxx envvars
+        lazy val consumedPorts = scala.collection.mutable.Set(definedPorts: _*) ++= assignedPorts.flatten
+        lazy val rand = new scala.util.Random(System.currentTimeMillis)
+        val maxPort: Int = 65535 - 1024
+
+        // carefully pick a container port that doesn't overlap with other ports used by this
+        // container. and avoid ports in the range (0 - 1024)
+        def next: Int = {
+          val p = rand.nextInt(maxPort) + 1025
+          if (!consumedPorts.contains(p)) {
+            consumedPorts += p
+            p
+          }
+          else next // TODO(jdef) **highly** unlikely, but still possible that the port range could be exhausted
+        }
+      }
 
       assignedPorts.zipWithIndex.foreach {
-        case (p, n) =>
+        // matches fixed or dynamic host port assignments
+        case (Some(p), n) =>
           env += (s"PORT$n" -> p.toString)
-      }
 
-      definedPorts.zip(assignedPorts).foreach {
-        case (defined, assigned) =>
-          if (defined != AppDefinition.RandomPortValue) {
-            env += (s"PORT_$defined" -> assigned.toString)
+        // matches container-port-only mappings; no host port was defined for this mapping
+        case (None, n) =>
+          definedPorts.lift(n) match {
+            case Some(containerPort) if containerPort == AppDefinition.RandomPortValue =>
+              val cp = ContainerPortGenerator.next
```

Comment:
Our documentation says:

> Using the value 0 for any port settings indicates to Marathon that you would like a random port assignment. However, if containerPort is set to 0 within a portMapping, it is set to the same value as hostPort.

Wouldn't this new code result in `hostPort != containerPort` if both are `0`?